### ISSUE TITLE
Shared folder "lost" after being renamed

### ISSUE THREAD

**Issue by Holt59** [nextcloud/server#21407]
### Steps to reproduce
1. User A creates folder "X" and shares it with a group containing user B.
2. User B renames folder "X" to "Y".

### Expected behaviour

I would expect either:
- The folder being renamed only for user B (as it is the case if user A renames it I think), or
- The folder being renamed for everyone, or
- An error saying that user B cannot rename the folder?

### Actual behaviour

The folder disappears for user B.

**Note:** If user A shares the folder to user B again (only to user B, did not test for a whole group), then user B get the folder back **with the new name**. Removing user B (but keeping the group containing user B) makes the folder disappear again for user B.

### Server configuration

**Operating system:** PhotonOS

**Web server / Database / PHP version:** The ones from the default docker configuration. 

**Nextcloud version:** 18.0.5

**Updated from an older Nextcloud/ownCloud or fresh install:** Yes.

**Where did you install Nextcloud from:** docker

**Signing status:**
<details>
<summary>Signing status</summary>

```
No errors have been found.
```
</details>

**Are you using external storage, if yes which one:** Yes (local folder), but not for the related folder.

**Are you using encryption:** no

**Are you using an external user-backend, if yes which one:** LDAP

### Client configuration
**Browser:** Firefox

**Operating system:** Windows

### Logs

Cannot access logs right now.

**Comment by Dragon988** [comment]
Have you verified this bug exists in 18.0.6? Noticed you were on 18.0.5.

**Comment by sharidas** [comment]
With version NC20, and steps shared in the issue, the move operation is successful for me. I can see a new entry in the share table (i.e, the updated name of the folder). The folders are visible for the sender and the recipient of the share.

**Comment by GitHubUser4234** [comment]
We found this bug in 18.0.7 as well. Renaming a shared a folder makes it disappear and the sharing person has to re-share again to make it re-appear. Definitely not a small bug I'd say...

**Comment by szaimen** [comment]
Based on sharidas's comment I am closing this since it seems to be fixed on NC20 and above.

**Comment by darknight7** [comment]
I have experienced this very same bug with two different users in NC 27 and 28. I've been able to reproduce it right now in NC 28.0.2 the only difference with the original report is that I don't use external storage. PHP FPM 8.2.15, redis 7.2.4, PHP opcaches enabled.

I've done a little investigation.

1. User A creates new folder TARAPIA and shares it with user B
2. User B renames TARAPIA to TERAPIA
3. The TERAPIA link is now un-clickable for B and the share disappears from the interface upon a reload
4. However if I visit the old url of the share as B two things happen:
5. - I get redirected to the new correct url
6. - the share now is listed in the interface
7. User B renames TERAPIA to TORAPIA -> back to point 3

I attach a log grep, anonimized.
